### PR TITLE
Add support for pretty printing Emscripten heap objects

### PR DESCRIPTION
**Pull request by hackcasual**
This uses the DI information extracted by the LLVM changes in the emscripten-fastcomp pull request.

By enabling CYBERDWARF, a user can dump a JSON formatted view of a variable.

test_pointers.cpp should give a good idea of how to use the initial debugger.

### CODE REVIEW

File: emcc.py [hunk 1]
```diff
@@ -1512,7 +1521,10 @@ def repl(m):
         if DEBUG:
           # Copy into temp dir as well, so can be run there too
           shared.safe_copy(memfile, os.path.join(shared.get_emscripten_temp_dir(), os.path.basename(memfile)))
-        return 'memoryInitializer = "%s";' % os.path.basename(memfile)
+        if shared.Settings.CYBERDWARF:
+          return 'memoryInitializer = "%s"; emDebugCDFile = "%s";' % (os.path.basename(memfile), os.path.basename(memfile)[:-3] + "cd")
```
Comment (hackcasual) [inline emcc.py]: This is quite ugly, and I'm not expecting it to ship this way. However I'm not sure of a good way to pass the name of the .cd file

File: tests/debugger/test_union.cpp [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    int i;
+    float f;
+};
+
+void test_1() {
+  TestBase tb;
+  tb.f = 13.37;
+
+  EM_ASM_INT({
+    var decoded = Module['emdebugger_heap_printer'].decode_from_stack($0, "tb", 100)["union TestBase"];
```
Comment (hackcasual) [inline tests/debugger/test_union.cpp]: No asserts, this test is still a WIP, though running it directly from a JS_ENGINE will give an idea of what the output looks like

File: emcc.py (hunk 1 above)
Comment (kripken) [inline emcc.py]: To pass it into the emitted JS? Define a variable in settings.js, set shared.Settings.THAT_VAR in emcc.py, and then you can use that variable either in emscripten.py (on settings) or in the JS files, for example, you can do `{{{ THAT_VAR }}}`, or `#if THAT_VAR` etc, in preamble.js.

File: emscripten.py [hunk 3]
```diff
@@ -1153,7 +1159,7 @@ def finalize_output(metadata, post, funcs_js, need_asyncify, provide_fround, asm
        access_quote('asmGlobalArg'), the_global,
        shared_array_buffer,
        access_quote('asmLibraryArg'), sending,
-       "'use asm';" if not metadata.get('hasInlineJS') and settings['ASM_JS'] == 1 else "'almost asm';", 
+       "'use asm';" if not metadata.get('hasInlineJS') and settings['ASM_JS'] == 1 and not settings['CYBERDWARF'] else "'almost asm';",
```
Comment (kripken) [inline emscripten.py]: why does this not allow asm validation?

File: src/library_debugger_toolkit.js [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+    "decode_from_stack": stack_decoder,
+    "install_cd_file": install_cyberdwarf,
+    "set_current_function": set_current_function,
+    "decode_var_by_var_name": pretty_print_to_object,
+    "decode_var_by_type_name": pretty_print_from_typename
+
+  };
+};
+
+mergeInto(LibraryManager.library, {"emdebug_Debugger": EMDebugHeapPrinter});
```
Comment (kripken) [inline src/library_debugger_toolkit.js]: space after `{` and before `}`

File: src/postamble.js [hunk 5]
```diff
@@ -94,6 +94,39 @@ if (memoryInitializer) {
 #endif
 #endif
 
+#if CYBERDWARF
```
Comment (kripken) [inline src/postamble.js]: looks like a copy-paste of the mem init code? perhaps we should consider a refactoring to share? not sure if feasible.

**Comment by kripken** [comment]
Where is the name CYBERDWARF from? :)

File: emscripten.py (hunk 3 above)
Comment (hackcasual) [inline emscripten.py]: I'll back this out. There's another change I'm working on that emits metadata intrinsics. There's code for it in CallHandlers.h but it's currently behind an option flag that's not plumbed through. In that case it inserts function calls when some source variable is assigned like:

`metadata_llvm_dbg_value_local($jsvar, type_id, offset, DWARF_OP, "source variable name")`

I believe that would not validate with asm.js. My thinking is that there would be additional configuration options for emcc to control intrinsics emitting.

**Comment by hackcasual** [comment]
CyberDWARF, like DWARF but for the web ;). It's also a character from the "Barkley, Shut Up and Jam: Gaiden" saga

**Comment by hackcasual** [comment]
Screwed up collapsing the work addressing the comments, opened up a cleaner PR here: https://github.com/kripken/emscripten/pull/4174

File: emscripten.py (hunk 3 above)
Comment (kripken) [inline emscripten.py]: I see. Yeah, that would not validate as asm.js, so it's correct to remove `use asm`.

**Pull request closed without merging**